### ISSUE TITLE
ENH: `df.explode` should turn empty lists into *no* rows rather than a NaN row

### ISSUE THREAD

**Issue by rben01** [pandas-dev/pandas#39915]
#### Is your feature request related to a problem?

When using `df.explode`, I'd expect that a list-like `L` gets transformed into `len(L)` rows and a scalar into 1 row. While documented, the current behavior of turning empty list-likes into 1 row containing NaN instead of 0 rows is counterintuitive.

#### Describe the solution you'd like

Ideally `df.explode` would transform empty list-likes into 0 rows, or, to maintain backwards compatibility, introduce an argument `collapse_empty=False` (or similar -- maybe `implode_empty` in keeping with the theme) that I could set to `True`.

#### API breaking implications

If `collapse_empty=False` by default then there would be no API breaking changes.

#### Describe alternatives you've considered

Carefully using `df.dropna` after `df.explode` to remove any null values I don't want while preserving null values that had belonged to lists prior to the explode.

#### Additional context

Using the example in the docs here: https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.explode.html

```python
>>> df = pd.DataFrame({'A': [[1, 2, 3], 'foo', [], [3, 4]], 'B': 1})
>>> df
           A  B
0  [1, 2, 3]  1
1        foo  1
2         []  1
3     [3, 4]  1

>>> df.explode('A', collapse_empty=True)
     A  B
0    1  1
0    2  1
0    3  1
1  foo  1
3    3  1
3    4  1

```

**Comment by rhshadrach** [comment]
It seems easier to subset before the explode, although I'm uncertain whether this will cover all cases:

```
df = pd.DataFrame({'A': [[1, 2, 3], 'foo', [], [3, 4], 4, None, np.nan, '', [np.nan]], 'B': 1})
mask = df['A'].str.len()
df = df[mask.isnull() | mask.gt(0) | df['A'].eq('')]
exploded = df.explode('A')
```

**Comment by jreback** [comment]
there is another issue about this - see if u can find / link

**Comment by lithomas1** [comment]
xref #35840 (@jreback was this what you were looking for?)

**Comment by rben01** [comment]
My issue is slightly different because I’m suggesting collapsing the exploded list instead of returning something other than NaN, but I can close this issue in favor of that one if you think the two issues are similar enough.

**Comment by rhshadrach** [comment]
I agree this is of a different nature - inferring dtypes one still can't determine if the resulting value was from an empty list or `np.nan` or `[np.nan]`.

However I do think that the right solution here is to filter empty lists before doing explode as this keeps the API small. Updated my comment above with a safer and more performant implementation.

**Comment by jreback** [comment]
#34923 as well

**Comment by rben01** [comment]
> I agree this is of a different nature - inferring dtypes one still can't determine if the resulting value was from an empty list or `np.nan` or `[np.nan]`.
> 
> However I do think that the right solution here is to filter empty lists before doing explode as this keeps the API small. Updated my comment above with a safer and more performant implementation. @rben01 - does that solution suit your needs?

Your implementation seems bugged -- `df.explode` doesn't check `isinstance(x, list)` but rather `c_is_list_like(x)`, which could lead to your filtering not catching empty list-likes. The difficulty in getting this *exactly* right is why I think it should be moved into the API.
